clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -77,19 +78,26 @@
     private boolean retry;
     private boolean reopened;
 
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long createdMs) {
+    public ProducerBatch(final TopicPartition tp, final MemoryRecordsBuilder recordsBuilder, final long createdMs) {
         this(tp, recordsBuilder, createdMs, false);
     }
 
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long createdMs, boolean isSplitBatch) {
+    public ProducerBatch(final TopicPartition tp, final MemoryRecordsBuilder recordsBuilder,
+                         final long createdMs, final boolean isSplitBatch) {
+        this(tp, recordsBuilder, createdMs, isSplitBatch, new ProduceRequestResult(tp), new ArrayList<>());
```

Comment:
I worry a slight amount about memory consumption here, as we will be allocating a new array with default capacity of length 10 for every record batch, whether a batch is split or not. I wonder if we could delay the creation of the array list until isSplitBatch is true? If not, maybe we should pre-size the arraylist - the downside there is we'd need a comment explaining that.

Reply:
Agree, we can check if the `List` has been created before adding the batches to it, I will update the code.

Reply:
Done